### ISSUE TITLE
Better email validations in Accounts 

### ISSUE THREAD

**Issue by allisonhowell** [openstax/bit#243]
### Description
We have seen a huge spike in users not getting their PIN number from accounts. 
We also have a high percentage of bounced emails in AmazonSES, which will cause more and more email failures if it's not resolved.

We'd like to add DNS validation - Accounts does not verify an email address before they send the email, so we have a large number of bounces, which downgrades our rating. This could be a reason that we seem to be going to spam or not reaching the inboxes of users at particular schools. 
We want to add a check against the MX records of the email domain and make sure they exist.

It seems reasonable that the most common email prefixes do not need their records checked. We should skip this validation for domains like gmail.com, hotmail.com, outlook.com, yahoo.com, icloud.com, rice.edu.

Example code
```
mx = Resolv::DNS.open { |dns| dns.getresources('domain.com',  Resolv::DNS::Resource::IN::MX) }
ok = mx.size > 0
```

### Acceptance Criteria

- Fewer users are able to signup with invalid email addresses.
- Amazon SES score improves (may take 30-45 days, per Scott)

### Checklist for Done
- 2-DDD
  - [x] Description is complete 
  - [ ] Acceptance Criteria is complete
  - [ ] Test criterias are defined
  - [ ] Estimate is updated
  - [ ] If the issue is a story, then Epic is selected
  - [ ] If the issue is included in the quarterly plan, then Milestone is selected
  - [ ] If the issue is included in a product phase, then Release is selected
- 2B-UX DESIGN
  - [ ] The UX design is complete and the design files are attached to the issue
  - [ ] The UX design is approved by the Product Manager
  - [ ] The test plan is drafted.
- 3-CODE
  - [ ] The code change meets the acceptance criteria (for a story or change request), or fixes the identified defect (for a bug)
  - [ ] Affected areas in the code are identified
  - [ ] The code change is merged to master (dev review done, test suite passed, code coverage maintained)
- 3A-NOT CODE
  - [ ] The work is completed and meets the acceptance criteria for the issue
- 4A-UX REVIEW
  - [ ] The implemented feature is accepted by the UX Designer and the Product Manager
  - [ ] The test plan is finalized
- 5A-FUNCT VER
  - [ ] All tests in test plan passed
- 5C-REGRESSION
  - [ ] Affected areas that were identified are passed

- **QA needs checklist:**
  - [ ] Acceptance Criteria are complete
  - [ ] Milestone should be tagged for PR's 
  - [ ] Milestone should be tagged for issues
  - [ ] Content update needed
  - [ ] Content update complete

- **Deployment needs**
  - [ ] Pre-deployment tasks
  - [ ] Data migrations 
  - [ ] Post-deployment tasks

**Comment by jpslav** [comment]
It'd be great to do an analysis of the bounce notifications currently in the staff@openstax.org inbox, so that we know which are the biggest problems.  Will give us a better expectation of how this DNS check will help us.

Also, we should profile the DNS check to see how long it takes and if it impacts the user experience.  Possible that we can also skip DNS checks for the 10 most common domains that our users have (might save us time).

**Comment by BryanEliDimas** [comment]
Sounds great, @jpslav.

@allisonhowell I don't know how BIT's pipeline works but I am now working on this.

Progress update
-----
Currently I'm trying to figure out where exactly I should add this validation in the codebase. I'm thinking that it should probably go right below/after wherever we check if the email address is a valid **format** or not and let it do whatever _that_ does—which I'm going to guess is: it displays some kind of message asking users to please enter a valid email address. If so, this should be a very straight-forward fix.

As for doing an analysis of the bounce notifications currently in the staff@openstax.org inbox, @jpslav, I'm thinking that perhaps I can and should use Amazon's SES API to query for these numbers and then (maybe?) generate a CSV.

**Comment by mwharrison** [comment]
Hey @BryanEliDimas - Glad to have you on the team for a bit (ha!).
No need to worry about doing the email analysis - I can do that. 

I think JP's idea of doing the whitelisting for the most common email domains (gmail.com, hotmail.com, etc.) is a great idea! I'm going to add that to the card description.

Let me know if you have any questions.

**Comment by BryanEliDimas** [comment]
@mwharrison I think JP's idea for whitelisting is slightly different than what I think you're suggesting. JP's idea would involve querying the DB for the 10 most popular email domains and what you're suggesting involves simply adding a constant with the top email providers' domain worldwide or country-wide. I could do both, or what should I do?

**Comment by jpslav** [comment]
Actually I was thinking of what Michael said more but with a bit of what
Bryan said. Use a constant list in code with gmail yahoo etc but also do
one query on prod to see if we have a bunch of some domain to add to the
list. Definitely do not do real time queries of dB to see popular domains.
Later if we want to we can Store verified domains in redis.

We also should make sure that the dns check passes for every confirmed
email address on prod. If not we need a way to let people say “I’m sure
this email is right”

**Comment by BryanEliDimas** [comment]
1. I will add a constant with the most popular email address domains, which will include some of the most popular ones in _our_ DB, not just the most popular globally. After querying our DB to find such email domains, I have a list of 13 items total.
2. I am thinking of whitelisting additional email domains by storing them in the DB—our Postgres one, not Redis as I don't really see any advantages of using Redis over PG.
3. If we don't do real time queries of DB to see popular domains, @jpslav, then we would have to make an http request to find out if the email domain has an MX record associated with it or not. It's a trade off and I'm pretty sure that the latter is significantly less efficient (to wait on an external network request) and more error-prone (could we get blocked from pinging too often looking for MX records?).
4. **re:** _We also should make sure that the dns check passes for every confirmed
email address on prod. If not we need a way to let people say “I’m sure
this email is right”._ 
~Glad you mention that, it hadn't crossed my mind. That's an interesting problem... perhaps I can skip the validation if the account/email address/`ContactInfo` has already been verified? Or... actually, we really shouldn't have any accounts with invalid email addresses in our DB in the first place, as any email we try to send will fail, so we could and should do a **purge**™️ before or when rolling out this MX validation feature. If any real users were using a fake or mistyped email address repeatedly and they really want to recover access with that email, then I think it's appropriate that they be forced to contact CSupport – or righteously get locked out.~ 
EDIT: My validation will only run when creating a new account or when updating it, and we really don't want to allow users to use a fake or mistyped email address because why would we want to intentionally allow that?
5. A big challenge I encountered is: after my change(s), **a lot** of the tests fail 😣, too many to fix because the email addresses are either hard-coded or automagically being generated, so to get around that I'm being forced to change my implementation entirely. I've come to the conclusion that what I really need is a global variable to define the validator/strategy for checking MX records, and a fake/mock to switch to before every single test. (Therefore, I'm thinking of adding my validator code in an initializer.)

**Comment by BryanEliDimas** [comment]
The hard part is DONE, including most tests. Here's my WIP branch: https://github.com/openstax/accounts/compare/validate_email_domain_mx

**Comment by BryanEliDimas** [comment]
My PR is ready for code review; I've requested Michael Harrison. https://github.com/openstax/accounts/pull/650

cc: @brittweinstein

**Comment by BryanEliDimas** [comment]
I've addressed all comments and suggestions on my PR after first code review. Requested code review again by JP and Michael Harrison.

**Comment by BryanEliDimas** [comment]
@allisonhowell My PR is ready to merge. Should be done by @mwharrison or @jpslav since I wrote the code.
cc: @brittweinstein

**Comment by Malar-Natarajan** [comment]
@BryanEliDimas 
Please see screenshot and video. If there is a typo in the domain, a PIN is still generated.
I tried gmai.com and rice.ed and they both generated the PIN


https://drive.google.com/file/d/1y587r0nDNUzQwY7nOFm9XZYMlfEGNNZ4/view?usp=sharing

![Screen Shot 2019-02-18 at 4.27.24 PM.png](https://images.zenhubusercontent.com/5b23e2cc835fb47f522b81ba/cc272201-66c5-419f-8a1c-dd9eac2333ee)

**Comment by BryanEliDimas** [comment]
Hey @Malar-Natarajan, here are a few email address domains you can test with...

All of the following should always 100% of the time pass because behind the scenes, they are whitelisted:
- gmail.com
- outlook.com
- yahoo.com
- icloud.com
- hotmail.com
- uga.edu
- vols.utk.edu
- email.vccs.edu
- ku.edu
- bruinmail.slcc.edu

The following should also pass work 100% of the time because we _know_ they are real:
- rice.edu
- openstax.org
- harvard.edu
- ibm.com
.... etc.

As for testing email addresses that should never work, you can try typing in random junk like letters with numbers like:
- a09sd68f7.com
- notevenarealdomain101.com
- this1looksrealbutitsnotlol.com

**Comment by BryanEliDimas** [comment]
Note: Because the PR for this issue meets the requirements, it does exactly what it's supposed to, but it does not solve the larger _business problem_, we've agreed to create a separate issue for that. Here is my suggestion:
- Add to our database's "blacklist" or _not-trusted-domains list_ technically valid but nefarious email providers like `gmai.com`, `hotma.com`, `mdanderso.org`.

**Comment by Malar-Natarajan** [comment]
Closing this card as per above comments. Creating a new card with the issue found above

**Comment by Malar-Natarajan** [comment]
Verified in staging
